src/Storages/Kafka/StorageKafka2.cpp

```diff
@@ -1238,8 +1323,21 @@ std::optional<size_t> StorageKafka2::streamFromConsumer(ConsumerAndAssignmentInf
             return;
         consumer_info.consumer->updateOffsets(consumer_info.topic_partitions);
     });
+
+    auto & keeper_to_use = *consumer_info.keeper;
+
+    // Change temporary locks every TMP_LOCKS_REFRESH_POLLS calls (or whenever permanent_locks_changed is set)
+    // This keeps batchy work from sitting too long on the same partitions and smoothly hands off load over time.
+    ++consumer_info.poll_count;
+    if (consumer_info.poll_count >= TMP_LOCKS_REFRESH_POLLS)
+    {
+        updateTemporaryLocks(keeper_to_use, consumer_info.consumer->getAllTopicPartitions(), consumer_info.tmp_locks);
+        consumer_info.poll_count = 0;
+    }
```

Comment:
Let's only change locks in `streamToViews`. I think in this function the only things should happen is we fetch some messages, then stream them to the views. Do not change locks, because that means we also have to update the offset in the consumer. Let's move this logic to `streamToViews`.